### PR TITLE
CCD-1157 Update SA with sa_subnets

### PR DESCRIPTION
**Pull request by nathan-clark**
https://tools.hmcts.net/jira/browse/CCD-1157

**Does this PR introduce a breaking change?** (check one with "x")

- [ ] Yes
- [x] No

### CODE REVIEW

File: storage-account.tf
Comment (naikajah) [inline storage-account.tf:16]: @nathan-clark where is this provider defined?

File: storage-account.tf
Comment (naikajah) [inline storage-account.tf:43]: where is this provider defined?

File: storage-account.tf
Comment (naikajah) [inline storage-account.tf:3]: why is this local defined here? This doesn't seem like a good use of locals to me.

File: storage-account.tf
Comment (naikajah) [inline storage-account.tf:17]: where is this local defined?

File: storage-account.tf
Comment (naikajah) [inline storage-account.tf:18]: where are these locals defined?

Comment (nathan-clark) [reply storage-account.tf:21]: yes I couldn't get sandbox working to validate, got it working now ;)

**Comment by nathan-clark** [comment]
From Plan in AAT:

~ default_action             = "Allow" -> "Deny
           ip_rules                   = []
         ~ virtual_network_subnet_ids = [
             + "/subscriptions/1baf5470-1c3e-40d3-a6f7-74bfbce4b348/resourceGroups/aks-infra-cftptl-intsvc-rg/providers/Microsoft.Network/virtualNetworks/core-cftptl-intsvc-vnet/subnets/aks-00",
             + "/subscriptions/1baf5470-1c3e-40d3-a6f7-74bfbce4b348/resourceGroups/aks-infra-cftptl-intsvc-rg/providers/Microsoft.Network/virtualNetworks/core-cftptl-intsvc-vnet/subnets/aks-01",
             + "/subscriptions/1baf5470-1c3e-40d3-a6f7-74bfbce4b348/resourceGroups/aks-infra-cftptl-intsvc-rg/providers/Microsoft.Network/virtualNetworks/core-cftptl-intsvc-vnet/subnets/iaas",
             + "/subscriptions/8b6ea922-0862-443e-af15-6056e1c9b9a4/resourceGroups/aks-infra-preview-rg/providers/Microsoft.Network/virtualNetworks/core-preview-vnet/subnets/aks-00",
             + "/subscriptions/8b6ea922-0862-443e-af15-6056e1c9b9a4/resourceGroups/aks-infra-preview-rg/providers/Microsoft.Network/virtualNetworks/core-preview-vnet/subnets/aks-01",
             + "/subscriptions/96c274ce-846d-4e48-89a7-d528432298a7/resourceGroups/aks-infra-aat-rg/providers/Microsoft.Network/virtualNetworks/core-aat-vnet/subnets/aks-00",
             + "/subscriptions/96c274ce-846d-4e48-89a7-d528432298a7/resourceGroups/aks-infra-aat-rg/providers/Microsoft.Network/virtualNetworks/core-aat-vnet/subnets/aks-01",

Assuming this is correct, and exhaustive.

**Comment by naikajah** [comment]
Is it possible to separate AAT and Prod from this PR and keep it separate from other environments?  So we can test the changes in lower environments first.

**Review by MSancaktutar — APPROVED** [review]
Looks good to me, but we need a second review on this.

File: storage-account.tf
Comment (timja) [inline storage-account.tf:51]: same approach can be done here as in jenkins

File: storage-account.tf
Comment (timja) [inline storage-account.tf:24]: would probably be better to extract a local var with the subnet names,

`local.jenkins_subnets = ["iaas", "aks-00", "aks-01"]`

and then use a `for_each` over the local

and then use something like https://www.terraform.io/docs/language/functions/concat.html to join the list you get

File: prod.tfvars
Comment (timja) [inline prod.tfvars:4]: remove this var in each file you changed, set by jenkins-library

File: variables.tf
Comment (timja) [inline variables.tf:78]: the jenkins library will set this for you no need to set in each environment

🚫 Blocker
```suggestion
variable "aks_subscription_id" {}
```

File: storage-account.tf
Comment (timja) [inline storage-account.tf:72]: I suspect this won't work outside of AAT, you should try this in demo / sandbox first

**Review by timja — CHANGES_REQUESTED** [review]

File: state.tf
Comment (timja) [inline state.tf:7]: applies throughout, string interpolation not needed if it's just a variable, it will be causing lots of warnings like this
```suggestion
  subscription_id = var.aks_subscription_id
```

File: state.tf
Comment (timja) [inline state.tf:20]: shouldn't this be:

```suggestion
  subscription_id = var.mgmt_subscription_id
```

File: storage-account.tf
Comment (timja) [inline storage-account.tf:74]: should probably be flagged for only running in aat
```suggestion
  provider            = azurerm.aks-preview
  count               = var.env == "aat" ? 1 : 0
```

File: storage-account.tf
Comment (timja) [inline storage-account.tf:80]:
```suggestion
  provider             = azurerm.aks-preview
  count                = var.env == "aat" ? 1 : 0
```

File: storage-account.tf
Comment (timja) [inline storage-account.tf:7]: this needs some conditional to only load this in aat I think.

File: state.tf
Comment (nathan-clark) [reply state.tf:7]: yep 84 of them

**Comment by MSancaktutar** [comment]
> Is it possible to separate AAT and Prod from this PR and keep it separate from other environments? So we can test the changes in lower environments first.

Indeed, this would be much better.

File: state.tf
Comment (timja) [inline state.tf:20]: wrong subscription id

File: state.tf
Comment (timja) [inline state.tf:14]: wrong id

File: storage-account.tf
Comment (nathan-clark) [reply storage-account.tf:7]: i thought this will `var.env == "aat" ? join(",", local.sa_aat_subnets) `

File: variables.tf
Comment (timja) [inline variables.tf:78]: this value needs to be set not really sure how you plan on doing this though, I guess default it to the DCD-CFTAPPS-DEV subscription ID and then override it in sandbox to be DCD-CFTAPPS-SBOX id?

File: storage-account.tf
Comment (timja) [reply storage-account.tf:7]: not sure if it'll work but you can try it

File: variables.tf
Comment (nathan-clark) [reply variables.tf:78]: yes as per original tfvars, was just validating in sbox

**Review by timja — APPROVED** [review]

**Review by danlysiak — APPROVED** [review]

**Pull request merged**
